### PR TITLE
[WIP] Implement PGD in PyTorch

### PR DESCRIPTION
**Pull request by michaelshiyu**
#1020

### CODE REVIEW

File: tutorials/future/torch/mnist_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+
+def main(_):
+  device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+  # Load training and test data
+  data = ld_mnist()
+
+  # Instantiate model, loss, and optimizer for training
+  net = LeNet5(1, 2)
```
Comment (david-berthelot) [inline tutorials/future/torch/mnist_tutorial.py]: I think we lost the code to use the GPU when available.
```
if device == 'cuda': 
  net = net.cuda()
```

File: tutorials/future/torch/mnist_tutorial.py [hunk 2]
```diff
@@ -29,149 +34,81 @@ def forward(self, x):
...
+  optimizer = torch.optim.Adam(net.parameters(), lr=1e-4, weight_decay=0)
+
+  # This will be called later to report accuracy on clean and adversarial examples
+  def eval():
+    net.eval()
+    total = 0
+    total_correct = 0
+    total_correct_fgm = 0
+    total_correct_pgd = 0
+    for x, y in data.test:
```
Comment (michaelshiyu) [inline tutorials/future/torch/mnist_tutorial.py]: If using GPU, should perhaps do

x, y = x.to(device), y.to(device)

Otherwise, PyTorch throws an error if the model is already on GPU.

File: tutorials/future/torch/mnist_tutorial.py [hunk 3]
```diff
@@ -29,149 +34,81 @@ def forward(self, x):
...
-      x_adv_pgd = pgd.generate(
-        x,
-        y=y,
-        eps=.3,
-        eps_iter=.01,
-        nb_iter=40,
-        ord=np.inf
-        )
+  for epoch in range(FLAGS.nb_epochs):
+    for __, (x, y) in enumerate(data.train):
```
Comment (michaelshiyu) [inline tutorials/future/torch/mnist_tutorial.py]: x, y = x.to(device), y.to(device)

File: tutorials/future/torch/mnist_tutorial.py (hunk 1 above)
Comment (michaelshiyu) [reply tutorials/future/torch/mnist_tutorial.py]: or just

net.to(device)

File: tutorials/future/torch/mnist_tutorial.py [hunk 4]
```diff
@@ -29,149 +34,81 @@ def forward(self, x):
...
+  # Load training and test data
+  data = ld_mnist()
+
+  # Instantiate model, loss, and optimizer for training
+  net = LeNet5(1, 2)
+  loss_fn = torch.nn.CrossEntropyLoss(reduction='mean')
+  optimizer = torch.optim.Adam(net.parameters(), lr=1e-4, weight_decay=0)
+
+  # This will be called later to report accuracy on clean and adversarial examples
+  def eval():
```
Comment (michaelshiyu) [inline tutorials/future/torch/mnist_tutorial.py]: Maybe put the rest of this method under

with torch.no_grad():

to prevent any gradient from being computed/saved

Comment (npapernot) [reply tutorials/future/torch/mnist_tutorial.py]: This makes the attack fail

Comment (michaelshiyu) [reply tutorials/future/torch/mnist_tutorial.py]: @npapernot sorry I missed the fact that you initiated the attacks in the eval method, which requires gradient computations.

File: cleverhans/utils_pytorch.py [hunk 5]
```diff
@@ -183,7 +183,8 @@ def optimize_linear(grad, eps, ord=np.inf):
 
     max_abs_grad, _ = torch.max(abs_grad.view(grad.size(0), -1), 1)
     max_mask = abs_grad.eq(max_abs_grad.view(ori_shape)).to(torch.float)
-    num_ties = torch.sum(max_mask, red_ind, keepdim=True)
```
Comment (michaelshiyu) [inline cleverhans/utils_pytorch.py]: @npapernot I know this has already been merged but out of curiosity, could you please let me know what was the test case that revealed a bug in the original implementation? I've been testing with a few hand-crafted grad matrices but couldn't find a bug in the original implementation, so I guess I must have missed some corner cases. Thanks.

File: cleverhans/utils_pytorch.py [hunk 6]
```diff
@@ -183,7 +183,8 @@ def optimize_linear(grad, eps, ord=np.inf):
...
+    for red_scalar in red_ind:
+      num_ties = torch.sum(num_ties, red_scalar, keepdim=True)
     optimal_perturbation = sign * max_mask / num_ties
```
Comment (michaelshiyu) [inline cleverhans/utils_pytorch.py]: @npapernot I know this has already been merged but out of curiosity, could you please let me know what was the test case that revealed a bug in the original implementation? I've been testing with a few hand-crafted grad matrices but couldn't find a bug in the original implementation, so I guess I must have missed some corner cases. Thanks.
